source/postprocess/heating_statistics.cc

```diff
@@ -0,0 +1,177 @@
+                                      in.temperature);
+                fe_values[this->introspection().extractors.pressure]
+                .get_function_values (this->get_solution(),
+                                      in.pressure);
+                fe_values[this->introspection().extractors.velocities]
+                .get_function_values (this->get_solution(),
+                                      in.velocity);
+                fe_values[this->introspection().extractors.pressure]
+                .get_function_gradients (this->get_solution(),
+                                         in.pressure_gradient);
+                for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
+                  fe_values[this->introspection().extractors.compositional_fields[c]]
+                  .get_function_values(this->get_solution(),
+                                       composition_values[c]);
+                for (unsigned int i=0; i<fe_values.n_quadrature_points; ++i)
+                  {
+                    for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
+                      in.composition[i][c] = composition_values[c][i];
+                  }
+
+                in.position = fe_values.get_quadrature_points();
+
+                this->get_material_model().evaluate(in, out);
+                (*heating_model)->evaluate(in, out, heating_model_outputs);
+
+                for (unsigned int q=0; q<n_q_points; ++q)
+                  {
+                    local_heating_integrals += heating_model_outputs.heating_source_terms[q]
+                                                        * out.densities[q] * fe_values.JxW(q);
+
+                    local_mass += out.densities[q] * fe_values.JxW(q);
+                  }
+              }
+
+          // compute the sum over all processors
+          std::vector<double> local_value;
+          std::vector<double> global_value;
+          local_value.push_back(local_heating_integrals);
+          local_value.push_back(local_mass);
+
+          Utilities::MPI::sum (local_value,
+                               this->get_mpi_communicator(),
+                               global_value);
+          const double individual_global_heating_integrals = global_value[0];
+          const double global_mass = global_value[1];
+
+          // finally produce something for the statistics file
+          const std::string name1("Average " + heating_model_names[index] + " rate (W/kg) ");
+          statistics.add_value (name1, individual_global_heating_integrals/global_mass);
+          // also make sure that the other columns filled by the this object
+          // all show up with sufficient accuracy and in scientific notation
+          statistics.set_precision (name1, 8);
+          statistics.set_scientific (name1, true);
+
+          // TODO:
+          // Total internal heating rate is not making sense in 2D at the moment,
+          // need to put a scale factor to transfer it to 3D.
```

Comment:
I assume you say this because you have in mind that 2d is a rotationally symmetric domain. But what if you consider a 2d domain as a slice through an infinitely extended domain in the third direction?

There are likely many places in ASPECT where it is ambiguous what we mean when we do 2d computations.